### PR TITLE
target/ppc: Implement xscmp{eq,ge,gt}qp and move related functions to decodetree

### PR DESCRIPTION
**Pull request by vcoracolombo**

### CODE REVIEW

File: target/ppc/fpu_helper.c [hunk 1]
```diff
@@ -2147,28 +2147,31 @@ VSX_MADD(xvnmaddsp, 4, float32, VsrW(i), NMADD_FLGS, 0, 0)
...
+ *   fld   - vsr_t field
  *   svxvc - set VXVC bit
+ *   sdwr  - second doubleword result
  */
-#define VSX_SCALAR_CMP_DP(op, cmp, exp, svxvc)                                \
+#define VSX_SCALAR_CMP(op, tp, cmp, fld, exp, svxvc, sdwr)                    \
 void helper_##op(CPUPPCState *env, ppc_vsr_t *xt,                             \
                  ppc_vsr_t *xa, ppc_vsr_t *xb)                                \
 {                                                                             \
     ppc_vsr_t t = *xt;                                                        \
```
Comment (mferst) [inline target/ppc/fpu_helper.c:2163]: It seems that we're not using `xa` or `xb` after we set `t`, so I guess we could set `xt` directly.....

File: target/ppc/fpu_helper.c [hunk 2]
```diff
@@ -2179,23 +2182,23 @@ void helper_##op(CPUPPCState *env, ppc_vsr_t *xt,                             \
     vex_flag = fpscr_ve && (vxvc_flag || vxsnan_flag);                        \
                                                                               \
     if (!vex_flag) {                                                          \
-        if (float64_##cmp(xb->VsrD(0), xa->VsrD(0),                           \
+        if (tp##_##cmp(xb->fld, xa->fld,                                      \
                           &env->fp_status) == exp) {                          \
-            t.VsrD(0) = -1;                                                   \
-            t.VsrD(1) = 0;                                                    \
+            t.f128 = make_float128(-1ULL, sdwr);                              \
```
Comment (mferst) [inline target/ppc/fpu_helper.c:2187]: .... or initialize `t` with zero, use `memset` here, and drop the `else` branch.

File: target/ppc/fpu_helper.c [hunk 3]
```diff
@@ -2179,23 +2182,23 @@ void helper_##op(CPUPPCState *env, ppc_vsr_t *xt,                             \
...
-VSX_SCALAR_CMP_DP(xscmpeqdp, eq, 1, 0)
-VSX_SCALAR_CMP_DP(xscmpgedp, le, 1, 1)
-VSX_SCALAR_CMP_DP(xscmpgtdp, lt, 1, 1)
-VSX_SCALAR_CMP_DP(xscmpnedp, eq, 0, 0)
+VSX_SCALAR_CMP(xscmpeqdp, float64, eq, VsrD(0), 1, 0, 0)
+VSX_SCALAR_CMP(xscmpgedp, float64, le, VsrD(0), 1, 1, 0)
+VSX_SCALAR_CMP(xscmpgtdp, float64, lt, VsrD(0), 1, 1, 0)
+VSX_SCALAR_CMP(xscmpeqqp, float128, eq, f128, 1, 0, -1)
+VSX_SCALAR_CMP(xscmpgeqp, float128, le, f128, 1, 1, -1)
+VSX_SCALAR_CMP(xscmpgtqp, float128, lt, f128, 1, 1, -1)
```
Comment (mferst) [inline target/ppc/fpu_helper.c:2201]: New/moved helpers should be uppercase.

File: target/ppc/insn32.decode [hunk 4]
```diff
@@ -427,3 +431,12 @@ XXSPLTW         111100 ..... ---.. ..... 010100100 . .  @XX2
...
 LXVKQ           111100 ..... 11111 ..... 0101101000 .   @X_uim5
+
+## VSX Comparison Instructions
+
+XSCMPEQDP       111100 ..... ..... ..... 00000011 ...   @XX3
+XSCMPGEDP       111100 ..... ..... ..... 00010011 ...   @XX3
+XSCMPGTDP       111100 ..... ..... ..... 00001011 ...   @XX3
+XSCMPEQQP       111111 ..... ..... ..... 0001000100 -   @VX
+XSCMPGEQP       111111 ..... ..... ..... 0011000100 -   @VX
+XSCMPGTQP       111111 ..... ..... ..... 0011100100 -   @VX
```
Comment (mferst) [inline target/ppc/insn32.decode:442]:
```suggestion
XSCMPEQQP       111111 ..... ..... ..... 0001000100 -   @X
XSCMPGEQP       111111 ..... ..... ..... 0011000100 -   @X
XSCMPGTQP       111111 ..... ..... ..... 0011100100 -   @X
```

Comment (vcoracolombo) [reply target/ppc/insn32.decode]: Fixed, thanks

File: target/ppc/fpu_helper.c (hunk 3 above)
Comment (vcoracolombo) [reply target/ppc/fpu_helper.c]: fixed, thanks!

File: target/ppc/fpu_helper.c (hunk 2 above)
Comment (vcoracolombo) [reply target/ppc/fpu_helper.c]: I went with this idea. Even tough getting rid of a variable is good, I think removing the extra parameter makes the code more readable and might be more interesting than the possible (which I think would be small) gain in performance.

Thank you again :)

**Review by mferst — APPROVED** [review]

**Comment by mferst** [comment]
Merged in upstream: https://lists.gnu.org/archive/html/qemu-ppc/2022-03/msg00002.html

**Pull request closed without merging**